vizro-core/src/vizro/models/_action/_action.py

```diff
@@ -59,6 +58,57 @@ def build(self):
             "action_finished": Output("action_finished", "data", allow_duplicate=True),
         }
 
+        action_components = _get_action_callback_mapping(
+            action_id=self.id, argument="components"  # type: ignore[arg-type]
+        )
+
+        return callback_inputs, callback_outputs, action_components
+
+    def _action_callback_function(self, **inputs: Dict[str, Any]) -> Dict[str, Any]:
+        logger.debug("=============== ACTION ===============")
+        logger.debug(f'Action ID: "{self.id}"')
+        logger.debug(f'Action name: "{self.function._function.__name__}"')
+        logger.debug(f"Action inputs: {inputs}")
+
+        # Invoking the action's function
+        return_value = self.function(**inputs) or {}
+
+        # Raising the custom exception if return value length doesn't match the number of outputs
+        return_value_len = (
+            1 if not hasattr(return_value, "__len__") or isinstance(return_value, str) else len(return_value)
+        )
+        outputs = list(ctx.outputs_grouping.keys())
+        outputs.remove("action_finished")
+        if len(outputs) != return_value_len:
+            raise ValueError(
+                f"Number of action's returned elements: {return_value_len} does not match the number"
+                f" of action's defined outputs: {len(outputs)}."
+            )
+
+        if isinstance(return_value, dict):
+            return {"action_finished": None, **return_value}
+
+        # If return_value is a single element, ensure return_value is a list
+        if not isinstance(return_value, (list, tuple)):
+            return_value = [return_value]
```

Comment:
Shall we change the order here? Otherwise this might easily break as this currently depends on a specific code sequence to be executed.

When would the return_value be a tuple? I thought this would only return a list or dict?

```suggestion
        # If return_value is a single element, ensure return_value is a list
        if not isinstance(return_value, (list, tuple, dict)):
            return_value = [return_value]
        if isinstance(return_value, dict):
            return {"action_finished": None, **return_value}
        return {"action_finished": None, **dict(zip(outputs, return_value))}
```

Reply:
> When would the return_value be a tuple? I thought this would only return a list or dict?

Here is an example of the function that returns a tuple, we also want to support:
```
def fun():
    return 1, 2
```

> Maybe we could even push this further, to make it check for an iterable?

I tried, but string is also iterable object and we need to treat strings as single objects and convert them into a list.

Reply:
Let's continue this conversation [here](https://github.com/mckinsey/vizro/pull/80#discussion_r1348562316)